prov/verbs/src/verbs_ep.c

```diff
@@ -70,35 +101,47 @@ ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
 	return -FI_EAGAIN;
 }
 
-ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr)
+ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr, uint64_t flags)
 {
 	struct vrb_context *ctx;
+	struct vrb_domain *domain;
 	struct vrb_cq *cq;
 	struct ibv_send_wr *bad_wr;
 	struct ibv_wc wc;
+	size_t credits_to_give = 0;
 	int ret;
 
 	cq = container_of(ep->util_ep.tx_cq, struct vrb_cq, util_cq);
+	domain = vrb_ep_to_domain(ep);
 	cq->util_cq.cq_fastlock_acquire(&cq->util_cq.cq_lock);
 	ctx = ofi_buf_alloc(cq->ctx_pool);
 	if (!ctx)
 		goto unlock;
 
-	if (!cq->credits || !ep->tx_credits) {
+	if (!cq->credits || !ep->sq_credits || !ep->peer_rq_credits) {
 		ret = vrb_poll_cq(cq, &wc);
 		if (ret > 0)
 			vrb_save_wc(cq, &wc);
 
-		if (!cq->credits || !ep->tx_credits)
+		if (!cq->credits || !ep->sq_credits || !ep->peer_rq_credits) {
 			goto freebuf;
+		}
+	}
+
+	if ((wr->opcode == IBV_WR_SEND || wr->opcode == IBV_WR_SEND_WITH_IMM ||
+	     wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM) &&
```

Comment:
A macro or static inline would make this a easier to read -- e.g. vrb_wr_consumes_recv()